### PR TITLE
feat:Added support for listing secrets #348

### LINKED ISSUE #348: Ability to list all secret path names from the SecretProvider
# 🚀 Feature Request

### Relevant Package [**REQUIRED**]

go-mod-bootstrap SecretProvider

### Description [**REQUIRED**]
Not sure if this would introduce a secerity concern (though it only removes a small amount of security through obscurity which isn't really security anyways).

It would be nice to be able to list the path names for all secrets in the SecretStore. The target use case I envision would be the ability to do something like this:

- User can assign the secret credentials for an Onvif camera by being given a list of all of the existing secret names/paths and selecting which one they want to assign to it.

This is currently not possible because you cannot get a list of all of the secret names/paths, you need to know the name/path beforehand.

### Describe the solution you'd like
Something like this (name is just a suggestion):
```go
type SecretProvider interface {
	...

	// ListAllSecretPaths returns the top level name/path for all secrets contained in the SecretStore.
	ListAllSecretPaths() ([]string, error)
}
```

### Describe alternatives you've considered
Not really any alternatives or workarounds.

### PR DESCRIPTION
**Pull request by drkfmorton**
Signed-off-by: Kyle Morton <kylex.f.morton@intel.com>

Fixes #348

If your build fails due to your commit message not passing the build checks, please review the guidelines here: https://github.com/edgexfoundry/go-mod-bootstrap/blob/main/.github/Contributing.md

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] I am not introducing a breaking change (if you are, flag in conventional commit message with `BREAKING CHANGE:` describing the break)
- [x] I am not introducing a new dependency (add notes below if you are)
- [x] I have added unit tests for the new feature or bug fix (if not, why?)
- [x] I have fully tested (add details below) this the new feature or bug fix (if not, why?)
- [x] I have opened a PR for the related docs change (if not, why?)
  <link to docs PR>

### CODE REVIEW

**Comment by drkfmorton** [comment]
Starting this PR while I finish testing

File: bootstrap/interfaces/secret.go [hunk 1]
```diff
@@ -23,6 +23,9 @@ type SecretProvider interface {
 	// Service key is use as the access token role which must have be previously setup.
 	GetAccessToken(tokenType string, serviceKey string) (string, error)
 
+	// ListSecretsAtPath returns a list of secret keys from an insecure/secure secrets secret store.
+	ListSecretsAtPath(path string) ([]string, error)
```
Comment (lenny-intel) [inline bootstrap/interfaces/secret.go:27]: This is not what the enhancement issue is requesting. Should be
```suggestion
	// ListSecretPaths returns a list of secret paths from the service's SecretStore.
	ListSecretPaths() ([]string, error)
```

Comment (bnevis-i) [reply bootstrap/interfaces/secret.go:27]: @lenny-intel Nothing in either the Consul nor Vault implementation limits the secret store to being a flat directory.  Both of them allow for additional hierarchy to be interposed between the prefix and the actual secret.

Comment (bnevis-i) [reply bootstrap/interfaces/secret.go:27]: Update: Lenny's comment here https://github.com/edgexfoundry/go-mod-bootstrap/issues/348#issuecomment-1239664562 directly addreses the above.  Basically, Lenny is advocating that we remove support for hierarchy.  I am ok with that.

**Comment by drkfmorton** [comment]
Fixes #348

File: bootstrap/interfaces/secret.go [hunk 2]
```diff
@@ -23,6 +23,9 @@ type SecretProvider interface {
 	// Service key is use as the access token role which must have be previously setup.
 	GetAccessToken(tokenType string, serviceKey string) (string, error)
 
+	// ListSecretPaths returns a list of secret keys from an insecure/secure secrets secret store.
+	ListSecretPaths(path string) ([]string, error)
```
Comment (lenny-intel) [inline bootstrap/interfaces/secret.go:27]: This should be retuning the list of paths for the current service, not secret keys at a path

```suggestion
	ListSecretPaths() ([]string, error)
```

**Review by lenny-intel — CHANGES_REQUESTED** [review]

**Review by bnevis-i — APPROVED** [review]

**Comment by drkfmorton** [comment]
Need to update go.mod with the latest changes from go-mod-secrets.

File: bootstrap/secret/insecure_test.go [hunk 3]
```diff
@@ -94,6 +96,53 @@ func TestInsecureProvider_GetAccessToken(t *testing.T) {
...
 
+func TestInsecureProvider_ListPaths(t *testing.T) {
+	configAllSecrets := TestConfig{
+		InsecureSecrets: map[string]bootstrapConfig.InsecureSecretsInfo{
+			"REDIS": {
+				Path:    expectedPath,
+				Secrets: expectedSecrets,
+			},
+			"KONG": {
+				Path:    "/kongdb",
```
Comment (lenny-intel) [inline bootstrap/secret/insecure_test.go:107]:
```suggestion
				Path:    "kongdb",
```

File: bootstrap/secret/insecure_test.go [hunk 4]
```diff
@@ -94,6 +96,53 @@ func TestInsecureProvider_GetAccessToken(t *testing.T) {
...
+	}
+
+	tests := []struct {
+		Name         string
+		ExpectedKeys []string
+		Config       TestConfig
+		ExpectError  bool
+	}{
+		{"Valid", expectedSecretsKeys, configAllSecrets, false},
+		{"Invalid - No secrets", []string{"/redisdb"}, configMissingSecrets, false},
```
Comment (lenny-intel) [inline bootstrap/secret/insecure_test.go:128]:
```suggestion
		{"Invalid - No secrets", []string{"redisdb"}, configMissingSecrets, false},
```

File: bootstrap/secret/secure_test.go [hunk 5]
```diff
@@ -388,3 +388,34 @@ func TestSecureProvider_HasSecrets(t *testing.T) {
 		})
 	}
 }
+
+func TestSecureProvider_ListSecretPathsSecrets(t *testing.T) {
+	expectedKeys := []string{"username", "password", "config/"}
```
Comment (lenny-intel) [inline bootstrap/secret/secure_test.go:393]:
```suggestion
	expectedKeys := []string{"username", "password", "config"}
```

**Review by lenny-intel — CHANGES_REQUESTED** [review]

**Review by lenny-intel — APPROVED** [review]
LGTM

**Pull request merged**